File: api/routers/reports.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from core.database import get_db
from core.roles import Role
from dependencies.auth import has_role
from models.usuario import Usuario
from schemas.report import (
    ActiveUsersByActivityReportResponse,
    BillingReportResponse,
    ClassCancellationsReportResponse,
    ReportClassOption,
    UserRegistrationsReportResponse,
)
from services.report_service import ReportService

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/clients-registrations", response_model=UserRegistrationsReportResponse)
async def get_client_registrations_report(
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_client_registrations_report()


@router.get("/staff-registrations", response_model=UserRegistrationsReportResponse)
async def get_staff_registrations_report(
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_staff_registrations_report()


@router.get("/deleted-users", response_model=UserRegistrationsReportResponse)
async def get_deleted_users_report(
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_deleted_users_report()


@router.get("/active-users-by-activity", response_model=ActiveUsersByActivityReportResponse)
async def get_active_users_by_activity_report(
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_active_users_by_activity_report()


@router.get("/billing", response_model=BillingReportResponse)
async def get_billing_report(
    start_date: date,
    end_date: date,
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_billing_report(start_date, end_date)


@router.get("/classes", response_model=list[ReportClassOption])
async def list_report_classes(
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).list_report_classes()


@router.get("/class-cancellations/{clase_template_id}", response_model=ClassCancellationsReportResponse)
async def get_class_cancellations_report(
    clase_template_id: UUID,
    db: AsyncSession = Depends(get_db),
    _: Usuario = Depends(has_role(Role.ROLE_ADMIN)),
):
    return await ReportService(db).get_class_cancellations_report(clase_template_id)

File: frontend/src/actions/reports.ts
"use server";

import { serverApi } from "@/lib/server-api";
import {
    ActiveUsersByActivityReport,
    BillingReport,
    ClassCancellationsReport,
    ReportClassOption,
    UserRegistrationsReport,
} from "@/types/api";

function periodParams(startDate?: string, endDate?: string): Record<string, string> {
    return {
        ...(startDate ? { start_date: startDate } : {}),
        ...(endDate ? { end_date: endDate } : {}),
    };
}

async function getRegistrationsReport(
    endpoint: string,
    startDate?: string,
    endDate?: string,
): Promise<UserRegistrationsReport | null> {
    try {
        const res = await serverApi(endpoint, { params: periodParams(startDate, endDate) });
        if (!res.ok) return null;
        return res.json();
    } catch (error) {
        console.error("Get Registrations Report Error:", error);
        return null;
    }
}

export async function getClientRegistrationsReport(
    startDate?: string,
    endDate?: string,
): Promise<UserRegistrationsReport | null> {
    return getRegistrationsReport("/reports/clients-registrations", startDate, endDate);
}

export async function getStaffRegistrationsReport(
    startDate?: string,
    endDate?: string,
): Promise<UserRegistrationsReport | null> {
    return getRegistrationsReport("/reports/staff-registrations", startDate, endDate);
}

export async function getDeletedUsersReport(
    startDate?: string,
    endDate?: string,
): Promise<UserRegistrationsReport | null> {
    return getRegistrationsReport("/reports/deleted-users", startDate, endDate);
}

export async function getActiveUsersByActivityReport(
    startDate?: string,
    endDate?: string,
): Promise<ActiveUsersByActivityReport | null> {
    try {
        const res = await serverApi("/reports/active-users-by-activity", {
            params: periodParams(startDate, endDate),
        });
        if (!res.ok) return null;
        return res.json();
    } catch (error) {
        console.error("Get Active Users By Activity Report Error:", error);
        return null;
    }
}

export async function getBillingReport(
    startDate: string,
    endDate: string,
): Promise<{ data?: BillingReport; error?: string }> {
    try {
        const res = await serverApi("/reports/billing", {
            params: {
                start_date: startDate,
                end_date: endDate,
            },
        });

        if (!res.ok) {
            const errorData = await res.json().catch(() => ({}));
            return { error: errorData.detail || "No se pudo cargar el reporte." };
        }

        return { data: await res.json() };
    } catch (error) {
        console.error("Get Billing Report Error:", error);
        return { error: "No se pudo cargar el reporte." };
    }
}

export async function getReportClasses(): Promise<ReportClassOption[]> {
    try {
        const res = await serverApi("/reports/classes");
        if (!res.ok) return [];
        return res.json();
    } catch (error) {
        console.error("Get Report Classes Error:", error);
        return [];
    }
}

export async function getClassCancellationsReport(
    classId: string,
    startDate?: string,
    endDate?: string,
): Promise<ClassCancellationsReport | null> {
    try {
        const res = await serverApi(`/reports/class-cancellations/${classId}`, {
            params: periodParams(startDate, endDate),
        });
        if (!res.ok) return null;
        return res.json();
    } catch (error) {
        console.error("Get Class Cancellations Report Error:", error);
        return null;
    }
}
